### PR TITLE
feat(map_based_prediction): output rich prediction paths based on lanelet map for crosswalk users

### PR DESCRIPTION
**Pull request by satoshi-ota**
## Description

Now, cars, motorcycles, and trucks (buses) have rich prediction paths based on lanelet map, but for other objects, they have only naïve linear prediction path based on estimated speeds and time horizon.

In this PR, I updated `map_based_prediction` logic in order to output map based prediction path for crosswalk users.

Related Link: https://github.com/autowarefoundation/autoware.universe/issues/920

https://user-images.githubusercontent.com/44889564/168963918-6626878f-5380-4aa2-92ae-2e60725bf98c.mp4

### Who is crosswalk users ?

- pedestrian
- bicycle

### Target object (who has additional rich prediction path)

Crosswalk users who satisfies at least one of the following conditions are considered to have the intention to cross the crosswalk, and `map_based_prediction` node outputs additional prediction path based on map and estimated object velocity.

- move toward the crosswalk
- stop near the crosswalk

<div align="center">
  
</div>

### Map base prediction path for crosswalk users

The `map_based_prediction` node outputs additional prediction path from the target object's position to a reachable crosswalk entry point. If the target object is outside the road, the number of prediction paths are one per crosswalk.

<div align="center">
  
</div>

If the target object is inside the road or crosswalk, the number of prediction paths are depend on whether object is moving or not. If the object is moving, it outputs a prediction path toward an exit that existed in the direction of movement, but if the object has stopped, it is impossible to infer which exit points the object want to go, so the node outputs prediction paths toward both side exit point.

<div align="center">
  
</div>

If there are some crosswalks around the target object, multiple map based prediction paths may be output. On the other hand, even if the object is moving toward a crosswalk, the predicted path across the road is not output.

<div align="center">
  
</div>

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by satoshi-ota** [comment]
test at odaiba

- pedestrian stops in crosswalk

rviz_screenshot_2022_07_01-15_39_19

- pedestrian stops in road

rviz_screenshot_2022_07_01-15_39_55

- pedestrian stops in sidewalk

rviz_screenshot_2022_07_01-15_46_08

- pedestrian walks away from crosswalk

https://user-images.githubusercontent.com/44889564/176840126-50bd3c6a-7f33-4e9b-a5ec-4dc9fb72c143.mp4

- pedestrian walks toward crosswalk

https://user-images.githubusercontent.com/44889564/176840139-8ccc2241-a1b0-40f4-829d-072f4ec7bb84.mp4

- pedestrian walks in road

https://user-images.githubusercontent.com/44889564/176840155-7e59e95b-caae-4e02-8625-d24a2f545ee0.mp4

- pedestrian walks in crosswalk

https://user-images.githubusercontent.com/44889564/176840168-56a4f675-60c3-4c35-8bd9-d5df0eef68ba.mp4

**Comment by satoshi-ota** [comment]
@yukkysaito Could you review this PR ? :bow:

File: perception/map_based_prediction/src/map_based_prediction_node.cpp [hunk 1]
```diff
@@ -176,19 +375,121 @@ void MapBasedPredictionNode::objectsCallback(const TrackedObjects::ConstSharedPt
...
-      transformed_object.classification.front().label != ObjectClassification::TRAILER &&
-      transformed_object.classification.front().label != ObjectClassification::MOTORCYCLE &&
-      transformed_object.classification.front().label != ObjectClassification::TRUCK) {
-      PredictedPath predicted_path =
-        path_generator_->generatePathForNonVehicleObject(transformed_object);
-      predicted_path.confidence = 1.0;
-
+      label != ObjectClassification::CAR && label != ObjectClassification::BUS &&
+      label != ObjectClassification::TRAILER && label != ObjectClassification::MOTORCYCLE &&
+      label != ObjectClassification::TRUCK) {
```
Comment (yukkysaito) [inline perception/map_based_prediction/src/map_based_prediction_node.cpp:382]: pls separate for each class.
example
```
if(
     label == ObjectClassification::CAR && label == ObjectClassification::BUS ) {
    foo;
} else if (label == ObjectClassification::PEDESTRIAN){
    foo;
} else  if (label == ObjectClassification::UNKNOWN){
    foo;
}
}
```

File: perception/map_based_prediction/config/map_based_prediction.param.yaml [hunk 2]
```diff
@@ -1,9 +1,9 @@
 /**:
   ros__parameters:
     enable_delay_compensation: true
-    prediction_time_horizon: 10.0 #[s]
+    prediction_time_horizon: 15.0 #[s]
```
[the remark this answers is not in the document]
Comment (satoshi-ota) [reply perception/map_based_prediction/config/map_based_prediction.param.yaml:4]: I've been experimenting with this values, but I also think 10 sec is enough for planning, as you said.
So, I restore it for original value :+1:

[the remark this answers is not in the document]
Comment (satoshi-ota) [reply perception/map_based_prediction/config/map_based_prediction.param.yaml]: @yukkysaito Fix in https://github.com/autowarefoundation/autoware.universe/pull/922/commits/b81cd6fad378a75c9269720f013431cac04d0f4b

File: perception/map_based_prediction/src/map_based_prediction_node.cpp (hunk 1 above)
Comment (satoshi-ota) [reply perception/map_based_prediction/src/map_based_prediction_node.cpp]: @yukkysaito OK :+1: I fixed in https://github.com/autowarefoundation/autoware.universe/pull/922/commits/3b43997a42fa831c7e3a29a281faed6662e44043.

**Comment by satoshi-ota** [comment]
@yukkysaito I added documentation of the prediction algorithms for crosswalk user in https://github.com/autowarefoundation/autoware.universe/pull/922/commits/fcf0f5f179687666d8fc2cc87fca54ba3302b123.

**Review by yukkysaito — APPROVED** [review]
LGTM

**Pull request merged**